### PR TITLE
ENH: make epsilon optional in optimize.approx_fprime.

### PR DESCRIPTION
**Pull request by anntzer**
#### Reference issue
N/A

#### What does this implement/fix?

Most functions in scipy.optimize which need to approximate a gradient
already default to an epsilon step of sqrt(finfo(float).eps); it makes
sense (and is a minor usability improvement) to use the same default in
approx_fprime as well.

#### Additional information
N/A

### CODE REVIEW

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -877,7 +877,7 @@ def approx_fprime(xk, f, epsilon, *args):
         Should take `xk` as first argument, other arguments to `f` can be
         supplied in ``*args``. Should return a scalar, the value of the
         function at `xk`.
-    epsilon : array_like
+    epsilon : array_like, optional
```
Comment (andyfaff) [inline scipy/optimize/optimize.py:880]: Should be {float, array_like}.

File: scipy/optimize/optimize.py [hunk 2]
```diff
@@ -865,7 +865,7 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
     return result
 
 
-def approx_fprime(xk, f, epsilon, *args):
+def approx_fprime(xk, f, epsilon=_epsilon, *args):
```
Comment (andyfaff) [inline scipy/optimize/optimize.py:868]: It won't be clear from the docstring what that value is, does it work if we put `np.sqrt(np.finfo(float).eps)` here?

Back compatibility is important. I find calling this kind of signature a little bit confusing:

```
>>> def approx_fprime(xk, f, epsilon=1.0, *args):
...     print(xk, f, epsilon, args)
...
>>> approx_fprime(1, 2, epsilon=1.0, 1)
  File "<stdin>", line 1
    approx_fprime(1, 2, epsilon=1.0, 1)
                                      ^
SyntaxError: positional argument follows keyword argument
>>> approx_fprime(1, 2, epsilon=0.5, 1)
  File "<stdin>", line 1
    approx_fprime(1, 2, epsilon=0.5, 1)
                                      ^
SyntaxError: positional argument follows keyword argument
>>> approx_fprime(1, 2, epsilon=0.5, *(1.0,))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: approx_fprime() got multiple values for argument 'epsilon'
>>> approx_fprime(1, 2, epsilon=0.5, *(1.0,2.0))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: approx_fprime() got multiple values for argument 'epsilon'
>>> approx_fprime(1, 2, 0.5, *(1.0,2.0))
1 2 0.5 (1.0, 2.0)
```

See https://stackoverflow.com/questions/58018545/why-does-python-allow-args-after-keyword-arguments

I'm not sure the user confusion will be worth the amended PR.

File: scipy/optimize/optimize.py (hunk 1 above)
Comment (anntzer) [reply scipy/optimize/optimize.py:880]: sure

File: scipy/optimize/optimize.py (hunk 2 above)
Comment (anntzer) [reply scipy/optimize/optimize.py:868]: Regarding documenting the default value: I went for the same wording as in check_grad.

Regarding the API: Actually this doesn't change the API; whether before or after this PR `epsilon` was effectively positional-only if you wanted to pass `args`:
```
In [1]: def f(eps, *args): print(eps, args)

In [2]: f(eps=1, 2, 3)
  File "<ipython-input-2-2c21b3b3c320>", line 1
    f(eps=1, 2, 3)
                 ^
SyntaxError: positional argument follows keyword argument


In [3]: f(eps=1, *(2, 3))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-fee0e799560d> in <module>
----> 1 f(eps=1, *(2, 3))

TypeError: f() got multiple values for argument 'eps'
```
(In other words, regardless of whether `epsilon` has a default value or not, if you want to pass `args`, then you always need to pass `epsilon` positionally, not by keyword.)
I agree the signature is actually quite bad, but again this PR doesn't change the situation (perhaps `epsilon` should instead have been keyword-only, so the signature would have been `approx_fprime(xk, f, *args, epsilon=_epsilon)`, but OTOH that would have been somewhat annoying to use when `epsilon` had no default, and is in any case a much bigger change compat-wise).

**Comment by ev-br** [comment]
Is not approx_fprime (soft-) deprecated in favor of approx_derivative?

**Comment by anntzer** [comment]
I don't know?  At least only approx_fprime shows up under https://docs.scipy.org/doc/scipy/reference/optimize.html#finite-difference-approximation.

**Comment by andyfaff** [comment]
`approx_fprime` is the only public function and it's not deprecated. `approx_derivative` has a lot more functionality, but is private. It's also a lot more complicated to use, and is not really all that user friendly at the moment.

**Pull request merged**